### PR TITLE
Fix GenericPrecompileRegistry #36 

### PR DESCRIPTION
**Pull request by Naveenkhasyap**
1. Removed `ActivePrecomiles` 
2. Made necessary changes in test file for `ActivePrecomiles` 

built the code using make file post making changes 
and ran test file as well results attached 
Kindly review the changes.

Screenshot 2024-05-29 at 12 56 29 PM
Screenshot 2024-05-29 at 1 06 46 PM

### CODE REVIEW

File: concrete/concrete.go [hunk 1]
```diff
@@ -80,7 +80,8 @@ type PrecompileMap = map[common.Address]Precompile
 type PrecompileRegistry interface {
 	Precompile(address common.Address, blockNumber uint64) (Precompile, bool)
 	Precompiles(blockNumber uint64) PrecompileMap
-	ActivePrecompiles(blockNumber uint64) []common.Address
+	AddPrecompiles(startingBlock uint64, precompiles PrecompileMap)
+	AddPrecompile(startingBlock uint64, address common.Address, precompile Precompile)
```
Comment (therealbytes) [inline concrete/concrete.go]: These setters should not be in the interface. The interface and `GenericPrecompileRegistry` are separate things. The interface should only expose getters. But please do add a `PrecompiledAddresses` method that returns a slice of addresses of precompiles for the given block.

File: concrete/concrete_test.go
Comment (therealbytes) [inline concrete/concrete_test.go:1]: Please test the `PrecompiledAddresses` method as well (implement the method as described in my other comment). You can do it with something like this:

```golang
func verifyPrecompileSet(t *testing.T, registry *GenericPrecompileRegistry, num uint64, p pcSet) {
	// ...
	// Assert that PrecompiledAddresses returns the correct slice of addresses
	pcsAddr := registry.PrecompiledAddresses(num)
	expPcsAddr := make([]common.Address, 0, len(p.precompiles))
	for address := range p.precompiles {
		expPcsAddr = append(expPcsAddr, address)
	}
	r.ElementsMatch(expPcsAddr, pcsAddr)
}
```

File: concrete/concrete.go
Comment (therealbytes) [inline concrete/concrete.go:1]: This does not fix any of the issues in the `GenericPrecompileRegistry` implementation beyond removing the `ActivePrecompiles` method. For instance, consider how convoluted `index(blockNumber uint64) int` is. It should be:
```golang
func (c *GenericPrecompileRegistry) index(blockNumber uint64) int {
	for ii, startingBlock := range c.startingBlocks {
		if blockNumber < startingBlock {
			return ii - 1
		}
	}
	return len(c.startingBlocks) - 1
}
```

Note how `AddPrecompile` inserts precompiles and their addresses again even when there are already precompiles set for the passed block. The insert statements should be put inside an `} else {` clause.

Also, for consistency, please remove the redundant type arguments in calls to generic functions. `insert[uint64](...)` should become `insert(...)`.

**Review by therealbytes — CHANGES_REQUESTED** [review]

File: concrete/concrete.go (hunk 1 above)
Comment (Naveenkhasyap) [reply concrete/concrete.go]: Noted. will make necessary changes

File: concrete/concrete_test.go
Comment (Naveenkhasyap) [reply concrete/concrete_test.go:1]: Added in the test case

File: concrete/concrete.go
Comment (Naveenkhasyap) [reply concrete/concrete.go:1]: cleaned it up

**Comment by Naveenkhasyap** [comment]
Hi @therealbytes , updated the PR please review my chages.

File: concrete/concrete.go [hunk 2]
```diff
@@ -100,34 +100,30 @@ func NewRegistry() *GenericPrecompileRegistry {
 }
 
 func (c *GenericPrecompileRegistry) index(blockNumber uint64) int {
-	for ii, startingBlock := range c.startingBlocks {
+	// Iterate over blocks and return index if found else return -1
```
Comment (therealbytes) [inline concrete/concrete.go]: This comment is factually incorrect. Please fix it or remove it.

File: concrete/concrete.go [hunk 3]
```diff
@@ -100,34 +100,30 @@ func NewRegistry() *GenericPrecompileRegistry {
...
+	return len(c.startingBlocks) - 1
 }
 
 func (c *GenericPrecompileRegistry) AddPrecompiles(startingBlock uint64, precompiles PrecompileMap) {
 	idx := c.index(startingBlock)
 	if idx >= 0 && c.startingBlocks[idx] == startingBlock {
 		panic("precompiles already set for this block")
-	}
+	} else {
+		//Append the addresses and then insert
```
Comment (therealbytes) [inline concrete/concrete.go]: These insert statements should not be in an else block, the ones in `AddPrecompile` should. We are checking for an unexpected condition that causes a panic, not handling two expected possibilities.

File: concrete/concrete.go
Comment (therealbytes) [inline concrete/concrete.go:1]: Code comments:
- Should be properly formatted (there must be a space between `//` and the comment).
- Should explain the reasoning and intent behind the code, clarify complex logic, provide a high-level overview, and highlight important details or constraints. They should _not_ state the obvious, repeat what the code is doing, or be misleading.

File: concrete/concrete.go [hunk 4]
```diff
@@ -167,16 +165,19 @@ func (c *GenericPrecompileRegistry) Precompiles(blockNumber uint64) PrecompileMa
...
 	idx := c.index(blockNumber)
-
 	if idx < 0 {
 		return []common.Address{}
 	}
 	return c.addresses[idx]
 }
 
 func insert[T any](slice []T, index int, value T) []T {
+	//edge case
```
Comment (therealbytes) [inline concrete/concrete.go]: This comment is not useful and it is not properly formatted (missing space, starts in lowercase). Better to just remove it.

File: concrete/concrete_test.go [hunk 5]
```diff
@@ -127,9 +123,8 @@ func verifyPrecompileSet(t *testing.T, registry *GenericPrecompileRegistry, num
 
 func verifyPrecompileSingle(t *testing.T, registry *GenericPrecompileRegistry, num uint64, p pcSingle) {
 	r := require.New(t)
-	// Assert that all the provided addresses have been returned and all the returned
-	// addresses were provided
-	addresses := registry.ActivePrecompiles(num)
+	//Assert that addresses match
```
Comment (therealbytes) [inline concrete/concrete_test.go]: Format. See my other comments about comments.

**Review by therealbytes — CHANGES_REQUESTED** [review]
Almost ready! Needs some changes.

File: concrete/concrete.go (hunk 4 above)
Comment (Naveenkhasyap) [reply concrete/concrete.go]: Done

File: concrete/concrete.go
Comment (Naveenkhasyap) [reply concrete/concrete.go:1]: Noted

File: concrete/concrete.go (hunk 3 above)
Comment (Naveenkhasyap) [reply concrete/concrete.go]: Got it will make necessary changes

File: concrete/concrete.go (hunk 2 above)
Comment (Naveenkhasyap) [reply concrete/concrete.go]: agreed

**Comment by Naveenkhasyap** [comment]
Hi @therealbytes , updated the PR please review my changes.

**Review by therealbytes — APPROVED** [review]

**Pull request merged**